examples/mdetr/loss.py

```diff
@@ -0,0 +1,301 @@
+
+
+def construct_positive_map(
+    logits: Tensor,
+    target_tokens: List[List[List[int]]],
+    indices: List[Tuple[Tensor, Tensor]],
+    tokenized: BatchEncoding,
+):
+    # construct a map such that positive_map[k, i,j] = True iff query i is associated to token j in batch item k
+    # For efficency, the construction happens on CPU, then the whole matrix is transferred to GPU in one go.
+    positive_map = torch.zeros(logits.shape, dtype=torch.bool)
+    for i, ((idx_src, idx_tgt), tgt) in enumerate(zip(indices, target_tokens)):
+        cur_tokens = [tgt[j] for j in idx_tgt]
+        for j, tok_list in enumerate(cur_tokens):
+            for (beg, end) in tok_list:
+                beg_pos = char_to_token(tokenized, i, beg)
+                end_pos = char_to_token(tokenized, i, end - 1)
+
+                if beg_pos is None and end_pos is None:
+                    raise ValueError(
+                        "At least one of beg_pos and end_pos must not be None"
+                    )
+                positive_map[i, idx_src[j], beg_pos : end_pos + 1].fill_(True)
+    return positive_map.to(logits.device)
+
+
+def masked_dict_accuracy(
+    pred_dict: Optional[Dict[str, Tensor]] = None,
+    label_dict: Optional[Dict[str, Tensor]] = None,
+    mask_dict: Optional[Dict[str, Tensor]] = None,
+    answer_type_key: Optional[str] = "answer_type",
+):
+    accuracies = OrderedDict()
+    for k in pred_dict.keys():
+        if mask_dict is None or mask_dict[k] is None:
+            mask = torch.ones_like(pred_dict[k])
+        else:
+            mask = mask_dict[k]
+        accuracies[f"{k}_accuracy"] = (
+            (pred_dict[k][mask].argmax(-1) == label_dict[k][mask]).sum() / mask.sum()
+            if mask.any()
+            else torch.as_tensor(1.0, device=mask.device)
+        )
+    weighted = sum(
+        [
+            accuracies[f"{k}_accuracy"] * mask_dict[k].sum()
+            for k in pred_dict.keys()
+            if k != answer_type_key
+        ]
+    )
+    accuracies["answer_total_accuracy"] = (
+        accuracies[f"{answer_type_key}_accuracy"]
+        * weighted
+        / label_dict[answer_type_key].numel()
+    )
+    return accuracies
+
+
+def masked_dict_cross_entropy(
+    pred_dict: Optional[Dict[str, Tensor]] = None,
+    label_dict: Optional[Dict[str, Tensor]] = None,
+    mask_dict: Optional[Dict[str, Tensor]] = None,
+):
+    losses = OrderedDict()
+    if pred_dict.keys() != label_dict.keys():
+        raise ValueError("Keys of pred_dict and label_dict must match")
+    for k in pred_dict.keys():
+        if mask_dict is None or mask_dict[k] is None:
+            mask = torch.ones_like(pred_dict[k])
+        else:
+            mask = mask_dict[k]
+        norm_factor = mask.sum() if mask.any() else 1.0
+        losses[f"{k}_loss"] = (
+            F.cross_entropy(pred_dict[k], label_dict[k]).masked_fill(~mask, 0).sum()
+            / norm_factor
+        )
+
+    return losses
+
+
+class MDETRLoss(nn.Module):
```

Comment:
you might've answered this already in a similar PR, but what's the reason these losses aren't in core? is it because it's only used for VQA/GQA?

Reply:
Because it includes the contrastive alignment loss, which relies on some stuff from transformers.